### PR TITLE
feat!: Implement `RequestManagerTandem`, remove `add_request` from `RequestList`, accept any iterable in `RequestList` constructor

### PR DESCRIPTION
**Pull request by janbuchar**
> Tandem, or in tandem, is an arrangement in which two or more animals, machines, or people are lined up one behind another, all facing in the same direction.[[1]](https://en.wikipedia.org/wiki/Tandem#cite_note-OED-1) Tandem can also be used more generally to refer to any group of persons or objects working together, not necessarily in line.[[1]](https://en.wikipedia.org/wiki/Tandem#cite_note-OED-1)
(https://en.wikipedia.org/wiki/Tandem)

- Inspired by https://github.com/apify/crawlee/blob/4c95847d5cedd6514620ccab31d5b242ba76de80/packages/basic-crawler/src/internals/basic-crawler.ts#L1154-L1177 and related code in the same class
- In my opinion, it implements the feature more cleanly and without polluting `BasicCrawler` (...any further)
- The motivation for the feature is twofold:
    1. Apify Actor development - it is common that an Actor receives a `requestListSources` input from the user, which may be pretty complex (regexp-based extraction from remote URL lists), and which is usually parsed using `apify.RequestList.open`. At the same time, the Actor wants to use the built in `RequestQueue`.
    2. Sitemap parsing (#248) - similar to 1, but not coupled to the Apify platform - we want to read URLs from a sitemap in the background, but the URLs should go through the standard request queue

## Breaking changes

- `RequestList` does not support `.drop()`, `.reclaim_request()`, `.add_request()` and `add_requests_batched()` anymore
   - `RequestManagerTandem` with a `RequestQueue` should be used for this use case, `await list.to_tandem()` can be used as a shortcut
- The `RequestProvider` interface has been renamed to `RequestManager` and moved to the `crawlee.request_loaders` package
- `RequestList` has been moved to the `crawlee.request_loaders` package
- The `BasicCrawler.get_request_provider` method has been renamed to `BasicCrawler.get_request_manager` and it does not accept the `id` and `name` arguments anymore
- The `request_provider` parameter of `BasicCrawler.__init__` has been renamed to `request_manager`
 
## TODO
- [x] new tests
- [x] fix existing tests

### CODE REVIEW

File: src/crawlee/storages/_request_list.py
Comment (janbuchar) [inline src/crawlee/storages/_request_list.py:96]: I'm pretty positive that removing `add_request` from `RequestList` is a good idea - the main purpose of the class is handling the `requestListSources` input from Apify anyway. Currently it is just an inefficient version of `RequestQueue`.

However, we may want to make a wrapper for `RequestSource` that just raises an error in `add_request`, so that crawlers that do not enqueue anything by design are possible. Thoughts?

File: src/crawlee/storages/_request_source.py
Comment (Pijukatel) [inline src/crawlee/storages/_request_source.py:82]: I see that this was not changed, just moved. But why:
processed_requests.setdefault(processed_request.unique_key, processed_request)
and not just 
processed_requests[processed_request.unique_key]=processed_request
?

Does it mean, that the processed_request.**unique_key** is not so unique and can be repeated?

Comment (janbuchar) [reply src/crawlee/storages/_request_source.py:82]: It may not be unique in the `requests` argument, so we deduplicate it like this.

File: tests/unit/storages/test_request_source_tandem.py
Comment (Pijukatel) [inline tests/unit/storages/test_request_source_tandem.py:35]: Maybe add docstring about what are we testing here. To me it seems very similar to the previous test and I am not really sure what code change can break this test and not the previous one.

File: src/crawlee/basic_crawler/_basic_crawler.py
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:367]: Why are we removing the `id` and `name` args?

File: src/crawlee/storages/__init__.py
Comment (vdusek) [inline src/crawlee/storages/__init__.py:2]: I am also wondering about the structure, taking into account the new inheritance chain. Perhaps it could look something like this:

```
$ ls -1 src/crawlee/storages/
_base_storage.py
_dataset.py  
_key_value_store.py        
_request_queue.py
```

```
$ ls -1 src/crawlee/request_sources/
_request_list.py      
_request_source_tandem.py
_request_provider.py  
_request_source.py
```

Opinions?

File: src/crawlee/storages/_request_source.py
Comment (vdusek) [inline src/crawlee/storages/_request_source.py:28]: Request providers?

File: src/crawlee/storages/_request_source.py
Comment (vdusek) [inline src/crawlee/storages/_request_source.py:77]:
```suggestion
        """Transforms a list of request-like objects into a list of `Request` objects."""
```

File: src/crawlee/storages/_request_source.py
Comment (vdusek) [inline src/crawlee/storages/_request_source.py:1]: You are talking about the request queue in the docstrings (I'm talking about the whole file)

File: src/crawlee/storages/_request_source_tandem.py
Comment (vdusek) [inline src/crawlee/storages/_request_source_tandem.py:31]: I don't know, but the fact that we both inherit from `RequestProvider` and also accept it as an input argument suggests there might be a design issue. Isn't it?

File: src/crawlee/storages/_request_source_tandem.py
Comment (vdusek) [inline src/crawlee/storages/_request_source_tandem.py:36]: so it does not implement the interface as a whole

File: tests/unit/basic_crawler/test_basic_crawler.py
Comment (vdusek) [inline tests/unit/basic_crawler/test_basic_crawler.py:68]: cannot we use crawler's default request provider in this case?

File: tests/unit/basic_crawler/test_basic_crawler.py
Comment (vdusek) [inline tests/unit/basic_crawler/test_basic_crawler.py:81]: cannot we use crawler's default request provider in this case?

File: tests/unit/basic_crawler/test_basic_crawler.py
Comment (vdusek) [inline tests/unit/basic_crawler/test_basic_crawler.py:104]: Is it important for these tests to use explicit request provider? Can't we use just pass the requests to the `run`?

**Review by vdusek — COMMENTED** [review]
Also, could you please add this to the new section in the "Request storage" guide?

File: src/crawlee/basic_crawler/_basic_crawler.py
Comment (janbuchar) [reply src/crawlee/basic_crawler/_basic_crawler.py:367]: Because those only apply to `RequestQueue`, and `self._request_provider` may contain any `RequestProvider`. Also if you do `await crawler.get_request_provider(); await crawler.get_request_provider(name="Frankie")`, the second call won't return a Frankie, which is confusing.

Comment (janbuchar) [reply src/crawlee/basic_crawler/_basic_crawler.py:367]: We could add a `get_request_queue` helper, but that may be even more confusing.

File: src/crawlee/storages/__init__.py
Comment (janbuchar) [reply src/crawlee/storages/__init__.py:2]: The new `request_sources` structure looks fine, I could do that :)

File: tests/unit/basic_crawler/test_basic_crawler.py
Comment (janbuchar) [reply tests/unit/basic_crawler/test_basic_crawler.py:104]: It is important in one or two of those tests, I believe. I'll see if I can change it in the other cases.

File: src/crawlee/storages/_request_source_tandem.py
Comment (janbuchar) [reply src/crawlee/storages/_request_source_tandem.py:36]: Well, true. We could either remove `name` from the `RequestProvider` - it doesn't seem to be used anywhere - or we could just defer to the underlying request provider. I'm inclined towards 1.

File: src/crawlee/storages/_request_source_tandem.py
Comment (janbuchar) [reply src/crawlee/storages/_request_source_tandem.py:31]: Well, what the `RequestSourceTandem` does is that it takes a `RequestProvider`, duct-tapes a `RequestSource` on top of it and then acts as a `RequestProvider` itself. I'm sure it's a known design pattern with a name and everything :smile:

Comment (vdusek) [reply src/crawlee/storages/_request_source_tandem.py:31]: Yeah, the source of my concerns probably arise from the fact, that Python uses the same syntax for "implement an interface" and regular inheritance.

File: src/crawlee/basic_crawler/_basic_crawler.py
Comment (vdusek) [reply src/crawlee/basic_crawler/_basic_crawler.py:367]: Yeah, it makes sense. Either provide your own instance or you will get the default one (based on config).

File: src/crawlee/storages/_request_source_tandem.py
Comment (vdusek) [reply src/crawlee/storages/_request_source_tandem.py:36]: Agreed.

File: src/crawlee/storages/_request_list.py
Comment (janbuchar) [reply src/crawlee/storages/_request_list.py:96]: @vdusek a thumbs up is not enough here! :slightly_smiling_face: What part of the message do you agree with? If you think that a wrapper is a good idea, what would be a good name?

**Comment by janbuchar** [comment]
> Also, could you please add this to the new section in the "Request storage" guide?

Done.

@vdusek @Pijukatel this is ready for review again.

**Comment by B4nan** [comment]
don't forget to update the upgrading guide

**Review by Pijukatel — APPROVED** [review]

**Review by vdusek — COMMENTED** [review]
I'm starting to like it, good job. Just a few more things...

File: src/crawlee/request_sources/__init__.py
Comment (janbuchar) [reply src/crawlee/request_sources/__init__.py:11]: If we want to have `RequestSource` and `RequestSourceTandem` in the same package (and I believe we do), I can't think of any...

File: src/crawlee/request_sources/_request_list.py
Comment (janbuchar) [reply src/crawlee/request_sources/_request_list.py:21]: I wondered as well, then decided to wait until #99 is done :slightly_smiling_face:

Comment (vdusek) [reply src/crawlee/request_sources/_request_list.py:21]: okay

File: src/crawlee/request_sources/_request_source.py
Comment (janbuchar) [inline src/crawlee/request_sources/_request_source.py:53]: Is there a good reason to keep this one here or should we move it to the `RequestProvider`? Since it requires support for updates to the `Request`, I'd prefer to see it in the provider. It would also make implementing a custom `RequestSource` easier.

Comment (vdusek) [reply src/crawlee/request_sources/_request_source.py:53]: Hmm, that makes sense to some extent. On the other hand, I feel that `mark_request_as_handled` and `reclaim_request` should be at the same level. You either do one or the other when processing a request, correct? Or maybe both should be moved?

Comment (janbuchar) [reply src/crawlee/request_sources/_request_source.py:53]: Well, if you don't need to support retries, you can do just fine with only `mark_request_as_handled`, and such use case makes some sense.

Comment (janbuchar) [reply src/crawlee/request_sources/_request_source.py:53]: Fetching the next request and marking it as completed later makes it possible to implement a minimal request loader that can survive a migration, which will be a requirement for both the fully fledged `RequestList` and the `SitemapLoader` or whatever we decide to call it.

**Review by vdusek — COMMENTED** [review]
Great, thanks! Only a few nits...

And before merging, could you please:
- Update the PR title to match the new version.
- Mention the breaking changes in the upgrading guide.
- We will also need to update the [Request storage](https://crawlee.dev/python/docs/guides/request-storage) and [Result storage](https://crawlee.dev/python/docs/guides/result-storage) guides to reflect these changes. It doesn't have to happen now, but could you at least open an issue for that?

**Review by vdusek — APPROVED** [review]
LGTM

**Pull request merged**
